cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs

```diff
@@ -0,0 +1,741 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+module Cardano.CLI.Shelley.Run.Read
+  ( -- * Metadata
+    MetadataError(..)
+  , renderMetadataError
+  , readFileTxMetadata
+  , readTxMetadata
+
+    -- * Script
+  , ScriptWitnessError(..)
+  , renderScriptWitnessError
+  , readScriptDataOrFile
+  , readScriptWitness
+  , readScriptWitnessFiles
+  , readScriptWitnessFilesThruple
+  , ScriptDecodeError (..)
+  , deserialiseScriptInAnyLang
+  , readFileScriptInAnyLang
+
+  -- * Script data (datums and redeemers)
+  , ScriptDataError(..)
+  , readScriptDatumOrFile
+  , readScriptRedeemerOrFile
+  , renderScriptDataError
+
+  -- * Tx
+  , CddlError
+  , CddlTx(..)
+  , IncompleteTx(..)
+  , readFileTx
+  , readFileTxBody
+  , renderCddlError
+  , readCddlTx -- For testing purposes
+
+  -- * Tx witnesses
+  , ReadWitnessSigningDataError(..)
+  , renderReadWitnessSigningDataError
+  , SomeWitness(..)
+  , ByronOrShelleyWitness(..)
+  , ShelleyBootstrapWitnessSigningKeyData(..)
+  , CddlWitnessError(..)
+  , readFileTxKeyWitness
+  , readWitnessSigningData
+
+  -- * Required signer
+  , RequiredSignerError(..)
+  , categoriseSomeWitness
+  , readRequiredSigner
+  ) where
+
+import           Prelude
+
+import           Control.Monad.Trans.Except (ExceptT (..), runExceptT)
+import           Control.Monad.Trans.Except.Extra (firstExceptT, handleIOExceptT, hoistEither, left)
+import qualified Data.Aeson as Aeson
+import           Data.Bifunctor (first)
+import qualified Data.ByteString.Char8 as BS
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import qualified Data.List as List
+import qualified Data.Text as Text
+import           Data.Word
+
+
+import           Cardano.Api
+import           Cardano.Api.Shelley
+
+--TODO: do this nicely via the API too:
```

Comment:
Not sure what this TODO means.

Reply:
That's actually a comment from Duncan. I think he wanted to expose the CBOR stuff from ledger in a nicer way.